phoenicis-tools/src/main/java/org/phoenicis/tools/lnk/LnkParser.java

```diff
@@ -0,0 +1,156 @@
+package org.phoenicis.tools.lnk;
+
+import org.apache.commons.io.IOUtils;
+import org.phoenicis.configuration.security.Safe;
+import org.phoenicis.tools.files.FilesManipulator;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+
+/**
+ * This file parses .lnk files
+ * https://msdn.microsoft.com/en-us/library/dd871305.aspx
+ */
+@Safe
+public class LnkParser extends FilesManipulator {
+    private final static int LNK_HEADER_SIZE = 0x4C;
+
+    private final static int FILE_LOCATION_INFO_FLAG_OFFSET_OFFSET = 0x08;
+    private final static int BASENAME_OFFSET_OFFSET = 0x10;
+    private final static int NETWORK_VOLUME_TABLE_OFFSET_OFFSET = 0x14;
+    private final static int FINALNAME_OFFSET_OFFSET = 0x18;
+    private final static int SHARE_NAME_OFFSET_OFFSET = 0x08;
+
+    /**
+     * Parses a .lnk shortuct
+     *
+     * @param file The path to the file
+     * @return a {@link LnkFile}
+     * @throws IOException if any error occurs
+     */
+    public LnkFile parse(File file) throws IOException {
+        this.assertInDirectory(file);
+        try (final InputStream inputStream = new FileInputStream(file)) {
+            return parse(inputStream);
+        }
+    }
+
+    /**
+     * Parses a .lnk shortuct
+     *
+     * @param rawLnkShortcutByteArray The byte array content of the sortcut
+     * @return a {@link LnkFile}
+     */
+    public LnkFile parse(byte[] rawLnkShortcutByteArray) {
+        final byte dataFlags = rawLnkShortcutByteArray[20];
+
+        final LnkFileAttribute fileAttributes = fetchFilesAttributes(rawLnkShortcutByteArray);
+        final LnkData lnkData = fetchLnkData(rawLnkShortcutByteArray);
+
+        final boolean isDirectory = fileAttributes.hasDirMask();
+        final boolean hasArguments = lnkData.hasArguments();
+
+        final int fileStart = fetchFileStart(rawLnkShortcutByteArray, lnkData);
+        final byte[] rawLnkContentWithoutHeader = Arrays.copyOfRange(rawLnkShortcutByteArray, fileStart,
+                rawLnkShortcutByteArray.length - 1);
+
+        final boolean isLocal = this.isLnkLocal(rawLnkContentWithoutHeader);
+        final String fileName = parseLnkContent(rawLnkContentWithoutHeader, isLocal);
+
+        return new LnkFile(isDirectory, isLocal, fileName, hasArguments);
```

Comment:
Will it only be these few arguments? Otherwise a builder might make sense.

Reply:
So far yes. And this is a internal immutable class (as long as we do not support .lnk creation). Therefore, we should not have to instanciate from the outside world